File: runtimes/cpp/test/security_limits_test.cpp
#include "twilic/wire.hpp"
#include "twilic/v2.hpp"
#include "twilic/codec.hpp"
#include "twilic/errors.hpp"
#include <cassert>
#include <functional>
using namespace twilic;
static void rejects(const std::function<void()>& action) {
  bool rejected = false;
  try { action(); } catch (const TwilicError&) { rejected = true; }
  assert(rejected);
}
int main() {
  Buffer tiny{0}; Reader reader(tiny);
  reader.claim_output(100);
  rejects([&] { reader.claim_output(100); });
  rejects([&] { reader.read_exact(static_cast<size_t>(-1)); });
  Buffer nested(70, 0xa1); nested.push_back(0xc0);
  rejects([&] { decode_v2(nested); });
  decode_v2(Buffer{0xa0});
  Buffer runs; for (auto n : {1, 0, 100000}) encode_varuint(n, runs);
  Reader rle(runs);
  rejects([&] { decode_u64_vector(rle, VectorCodec::Rle); });
}

File: runtimes/dart/test/security_limits_test.dart
import 'dart:typed_data';
import 'package:test/test.dart';
import 'package:twilic/src/wire.dart';
import 'package:twilic/src/v2.dart';

void main() {
  test('cumulative budgets and invalid lengths', () {
    final reader = Reader(Uint8List(1));
    reader.claimOutput(100);
    expect(() => reader.claimOutput(100), throwsA(isA<Exception>()));
    expect(() => reader.readExact(-1), throwsA(isA<Exception>()));
  });
  test('depth rejection', () {
    expect(() => decodeV2(Uint8List.fromList([...List.filled(70, 0xa1), 0xc0])),
        throwsA(isA<Exception>()));
  });
}
